# Sử Dụng Pandas Profiling (Ydata Profiling) Trên Dữ Liệu Customer Personality Analysis

Notebook này thực hiện phân tích khám phá dữ liệu (EDA) tự động sử dụng ydata_profiling trên dataset Customer Personality Analysis từ Kaggle: [https://www.kaggle.com/datasets/imakash3011/customer-personality-analysis](https://www.kaggle.com/datasets/imakash3011/customer-personality-analysis).

Chúng ta sẽ:
1. Cài đặt ydata_profiling (trước đây là pandas-profiling).
2. Sử dụng công cụ để tạo báo cáo EDA tự động và lưu vào file HTML.
3. Tiến hành EDA dựa trên báo cáo HTML (`profile_output.html`).


In [1]:
# Cài đặt ydata_profiling nếu chưa có (chạy một lần)
!pip install ydata-profiling

   ---------------------------------------- 0.0/2.8 MB ? eta -:--:--
   ---------------------------------------- 2.8/2.8 MB 14.9 MB/s  0:00:00
   ---------------------------------------- 0.0/30.3 MB ? eta -:--:--
   ----- ---------------------------------- 3.9/30.3 MB 19.5 MB/s eta 0:00:02
   --------- ------------------------------ 7.3/30.3 MB 17.4 MB/s eta 0:00:02
   ------------- -------------------------- 10.2/30.3 MB 16.3 MB/s eta 0:00:02
   ----------------- ---------------------- 13.4/30.3 MB 16.1 MB/s eta 0:00:02
   --------------------- ------------------ 16.5/30.3 MB 16.3 MB/s eta 0:00:01
   ------------------------- -------------- 19.7/30.3 MB 15.5 MB/s eta 0:00:01
   ----------------------------- ---------- 22.3/30.3 MB 15.3 MB/s eta 0:00:01
   --------------------------------- ------ 25.2/30.3 MB 15.2 MB/s eta 0:00:01
   ------------------------------------ --- 28.0/30.3 MB 14.9 MB/s eta 0:00:01
   ---------------------------------------- 30.3/30.3 MB 14.4 MB/s  0:00:02
  

## 1. Cài Đặt Ydata Profiling

Chạy lệnh `!pip install ydata-profiling` để cài đặt. Sau khi cài đặt, import và sử dụng.

## 2. Sử Dụng Công Cụ Ydata Profiling

Đọc dữ liệu và tạo báo cáo EDA tự động, lưu vào file HTML.

In [2]:
import pandas as pd
from ydata_profiling import ProfileReport

# Đọc dữ liệu từ file CSV
marketing_data = pd.read_csv("./marketing_campaign.csv")

# Hiển thị thông tin cơ bản về dữ liệu
print('Số hàng và cột:', marketing_data.shape)
print('\n5 hàng đầu:')
display(marketing_data.head())

Số hàng và cột: (2240, 1)

5 hàng đầu:


,ID\tYear_Birth\tEducation\tMarital_Status\tIncome\tKidhome\tTeenhome\tDt_Customer\tRecency\tMntWines\tMntFruits\tMntMeatProducts\tMntFishProducts\tMntSweetProducts\tMntGoldProds\tNumDealsPurchases\tNumWebPurchases\tNumCatalogPurchases\tNumStorePurchases\tNumWebVisitsMonth\tAcceptedCmp3\tAcceptedCmp4\tAcceptedCmp5\tAcceptedCmp1\tAcceptedCmp2\tComplain\tZ_CostContact\tZ_Revenue\tResponse
0,5524\t1957\tGraduation\tSingle\t58138\t0\t0\t0...
1,2174\t1954\tGraduation\tSingle\t46344\t1\t1\t0...
2,4141\t1965\tGraduation\tTogether\t71613\t0\t0\...
3,6182\t1984\tGraduation\tTogether\t26646\t1\t0\...
4,5324\t1981\tPhD\tMarried\t58293\t1\t0\t19-01-2...


In [3]:
# Tạo báo cáo EDA tự động sử dụng ProfileReport
profile = ProfileReport(marketing_data, title="Customer Personality Analysis EDA Report")

# Lưu báo cáo vào file HTML
profile.to_file("./profile_output.html")

print('Báo cáo đã được lưu vào: ./profile_output.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

100%|██████████| 1/1 [00:00<00:00, 11.77it/s]
c:\Users\ASUS\AppData\Local\Programs\Python\Python312\Lib\site-packages\ydata_profiling\visualisation\utils.py:73: UserWarning: Glyph 9 (	) missing from font(s) Arial.
  plt.savefig(
c:\Users\ASUS\AppData\Local\Programs\Python\Python312\Lib\site-packages\ydata_profiling\visualisation\utils.py:73: UserWarning: Glyph 9 (	) missing from font(s) Arial.
  plt.savefig(


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

Báo cáo đã được lưu vào: ./profile_output.html


### Nhận Xét
- Báo cáo bao quát toàn bộ dataset (29 cột, 2240 hàng), phát hiện missing values (Income: 24 NA), outliers, và tương quan (e.g., Income vs. MntWines).
- Thời gian tạo báo cáo khoảng 10-20 giây tùy kích thước dữ liệu.

### Kết Luận
Ydata_profiling tiết kiệm thời gian so với EDA thủ công, cung cấp insights nhanh chóng về phân bố, tương quan.

## 3. Tiến Hành EDA Trên Trang Tập Tin Profile_Output.html

Mở file `./profile_output.html` trong trình duyệt để xem báo cáo tương tác. Dưới đây là tóm tắt các insights chính từ báo cáo.

#### Các Insights Chính Từ Báo Cáo:
- **Overview**: 2240 quan sát, 29 biến (20 numerical, 9 categorical). 1.07% missing values chủ yếu ở Income.
- **Variables**: 
  - Income: Phân bố lệch phải (mean ~52k, median ~51k), outliers cao (>100k).
  - Education: Phân loại (PhD ~9%, Master ~22%).
  - MntWines: Tương quan cao với Income (0.68).
- **Correlations**: Ma trận tương quan hiển thị mối liên hệ giữa chi tiêu (Mnt*) và Income.
- **Missing Values**: Income có 24 NA, có thể impute bằng median.
- **Interactions**: Biểu đồ scatter cho thấy clusters khách hàng theo chi tiêu và thu nhập.

### Nhận Xét
- Báo cáo tương tác cho phép zoom, filter, giúp khám phá sâu (e.g., histogram cho từng biến).
- Phát hiện vấn đề: High cardinality ở Dt_Customer (ngày khách hàng), cần xử lý.
- Ưu điểm: Tự động, toàn diện; nhược điểm: Không tùy chỉnh sâu.

### Kết Luận
EDA tự động qua ydata_profiling cung cấp cái nhìn toàn diện về dữ liệu khách hàng, hỗ trợ quyết định phân khúc (segmentation) dựa trên thu nhập và hành vi mua sắm.

## Kết Luận Tổng Quát

- Ydata_profiling là công cụ mạnh mẽ cho EDA nhanh chóng, tiết kiệm thời gian so với thủ công.
- Insights từ báo cáo: Dữ liệu có missing values cần xử lý, phân bố lệch ở Income, tương quan mạnh giữa thu nhập và chi tiêu.
- Gợi ý mở rộng: Sử dụng insights để xây dựng mô hình phân loại khách hàng hoặc dự đoán churn.

